### ISSUE TITLE
More modern cmake enhancement ideas

### ISSUE THREAD

**Issue by dcwhite** [SCIInstitute/SCIRun#2203]
* Use Artichoke to manage Superbuild settings/parameters.
* Use Python-cmake project to encapsulate Python versions on Mac and Linux (maybe Windows? it's lacking 3.8)
* Use GenerateExportHeader to replace SCISHARE usages
* Refactor Superbuild projects to use new target syntax
* Investigate generator expressions for compile flags and other things
* Exoprt configurations might be useful for Superbuild externals

**Comment by dcwhite** [comment]
**Analysis (2026-06-13)**

Modern CMake enhancement ideas: Artichoke for Superbuild params, GenerateExportHeader (replaces SCISHARE), target syntax for Superbuild, generator expressions, file(GET_RUNTIME_DEPENDENCIES). The GET_RUNTIME_DEPENDENCIES note is especially useful for packaging (replaces manual rpath fixup). These are good long-term CMake hygiene items, best tackled alongside #2197 and #2200.

**Comment by dcwhite** [comment]
**Analysis and sequencing (2026-08-05)**

Grounding each of the seven bullets in what the tree actually does today, then ordering them by what unblocks what. Facts below are from `master` at 35feca9.

## Current state, per bullet

**1. Artichoke.** [Superbuild.cmake:270](https://github.com/SCIInstitute/SCIRun/blob/master/Superbuild/Superbuild.cmake#L270) hand-maintains a ~100-line block of `"-DEigen_DIR:PATH=${Eigen_DIR}"` strings forwarded to the inner project. Every new option must be added there by hand, and omitting one fails silently — the inner build just uses a default. Artichoke's `mark_as_superbuild()` annotates each variable where it is declared instead.

**2. python-cmake-buildsystem.** `PythonExternal.cmake` has three unrelated code paths: Unix `./configure` with hand-written `LDFLAGS=-Wl,-rpath,'$ORIGIN/../lib'`; macOS adding an `--enable-framework` build plus a `tar` step to preserve framework symlinks; Windows driving `PCbuild\build.bat` with a second MSBuild step for Debug and a workaround for `aka.ms/nugetclidl` redirecting to a Bing search page. A CMake-native CPython collapses these to one path.

**3. GenerateExportHeader.** 101 hand-written `share.h` files, 2106 `SCISHARE` uses. Each `share.h` does `#undef SCISHARE` and redefines it from `BUILD_<Target>`, so the meaning of `SCISHARE` in a translation unit depends on which `share.h` was included last.

**4. Target syntax for Superbuild.** Externals are wired by convention, not targets: each `*External.cmake` sets a `<Name>_DIR` cache variable and consumers reconstruct paths from it. `ZlibExternal.cmake` is representative.

**5. Generator expressions.** `src/` has 38 directory-scoped `include_directories` calls and **zero** `target_include_directories`; 143 `add_definitions`; 8 total `$<>` uses across `src/CMakeLists.txt` and the whole Superbuild. Config-specific flags are gated on `if(CMAKE_BUILD_TYPE)`, which is empty at configure time under multi-config generators (VS, Ninja Multi-Config).

**6. Export configurations for externals.** There is no install stage at all — `ZlibExternal.cmake` sets `INSTALL_COMMAND ""` and exports the *build* directory as `Zlib_DIR`. Nothing produces a relocatable tree with `<Name>Config.cmake` files.

**7. `file(GET_RUNTIME_DEPENDENCIES)`.** Dependency resolution is hand-rolled today: `src/CMake/scripts/macdeploy.sh` pipes `otool -L` through `sed`/`awk` to generate `install_name_tool` commands, `src/Main/osx_bundle/CMakeLists.txt` adds more hardcoded `install_name_tool -change` calls for the Python framework, and Qt is handled by `macdeployqt` with no Linux or Windows equivalent.

## Order

The dependency structure is simpler than the list suggests: **6 unblocks 4**, and nothing else has hard prerequisites.

| # | Item | Status |
|---|---|---|
| 1 | **(7)** `GET_RUNTIME_DEPENDENCIES` for deployment | Unblocked. Highest value. |
| 2 | **(3)** `GenerateExportHeader` for SCISHARE | Unblocked. Cheap in its narrow form. |
| 3 | **(5)** Generator expressions for flags | Unblocked. Fixes a real multi-config bug. |
| 4 | **(6)** Export configs for externals | Already tracked as #2627 → #2631–2640, deferred behind the VTK merge. |
| 5 | **(4)** Target syntax for Superbuild | Blocked by #2627; overlaps #2197. |
| 6 | **(1)** Artichoke | Optional. Ergonomics only, adds a third-party dep. |
| 7 | **(2)** python-cmake-buildsystem | Optional. Third-party version-lag risk. |

Rows 4 and 5 above — original bullets (6) and (4) — are **not** getting new sub-issues, since #2627 and #2197 already cover them and duplicating would fragment the tracking. Sub-issues are being opened for the other five, in the order above.

Two notes on the ranking:

- **(7) goes first** because it is both the highest-value item and completely unblocked. One prerequisite, which is free: [src/CMakeLists.txt:28](https://github.com/SCIInstitute/SCIRun/blob/master/src/CMakeLists.txt#L28) still declares `cmake_minimum_required(VERSION 3.9 FATAL_ERROR)`, below the 3.16 this needs. Since the Superbuild already requires 3.20, nothing can build with 3.9 today — the bump costs nothing.
- **(6) is the keystone but is correctly deferred.** The missing install stage is also the root cause of both failed Superbuild-caching attempts (#2513, #2576) — you cannot cache a tree that was never packaged. It stays behind VTK.

Correction to the earlier 2026-06-13 comment: it suggests tackling these alongside #2197 and #2200, without noting that #2197 is itself deferred behind the VTK branch. The bullet it pairs with (4) inherits that deferral rather than being ready to start.
